File: src/fastapi/tests/fixtures/test_accounts.py
"""
Permanent Test Accounts for Automation Testing

This module provides constants and helpers for the permanent test accounts
created by migration 20260303_create_permanent_test_accounts.sql.

These accounts are deterministic and will always exist after running migrations.
Use them for integration tests, e2e tests, and automation.
"""

from __future__ import annotations

# ==========================================
# TEST ACCOUNT CONSTANTS
# ==========================================

# Test User Account (Regular User)
TEST_USER_ID = "a1111111-1111-1111-1111-111111111111"
TEST_USER_EMAIL = "test.user@hanachan.test"
TEST_USER_PASSWORD = "TestPassword123!"
TEST_USER_DISPLAY_NAME = "Test User"

# Test Admin Account (Super Admin)
TEST_ADMIN_ID = "b2222222-2222-2222-2222-222222222222"
TEST_ADMIN_EMAIL = "test.admin@hanachan.test"
TEST_ADMIN_PASSWORD = "AdminPassword123!"
TEST_ADMIN_DISPLAY_NAME = "Test Admin"
TEST_ADMIN_ROLE = "super_admin"

# ==========================================
# ACCOUNT DATA DICTIONARIES
# ==========================================

TEST_USER = {
    "id": TEST_USER_ID,
    "email": TEST_USER_EMAIL,
    "password": TEST_USER_PASSWORD,
    "display_name": TEST_USER_DISPLAY_NAME,
}

TEST_ADMIN = {
    "id": TEST_ADMIN_ID,
    "email": TEST_ADMIN_EMAIL,
    "password": TEST_ADMIN_PASSWORD,
    "display_name": TEST_ADMIN_DISPLAY_NAME,
    "role": TEST_ADMIN_ROLE,
}

ALL_TEST_ACCOUNTS = [TEST_USER, TEST_ADMIN]


# ==========================================
# JWT TOKEN HELPERS
# ==========================================


def make_test_token_for_user(user_id: str, role: str = "authenticated") -> str:
    """
    Generate a test JWT token for a specific user.

    Args:
        user_id: The user's UUID
        role: The user's role (default: "authenticated")

    Returns:
        A test JWT token string
    """
    import base64
    import json

    header = (
        base64.urlsafe_b64encode(json.dumps({"alg": "HS256", "typ": "JWT"}).encode())
        .rstrip(b"=")
        .decode()
    )

    payload = (
        base64.urlsafe_b64encode(
            json.dumps({"sub": user_id, "role": role, "aud": "authenticated"}).encode()
        )
        .rstrip(b"=")
        .decode()
    )

    return f"{header}.{payload}.test-signature"


def get_test_user_headers() -> dict:
    """Get authorization headers for the test user."""
    return {"Authorization": f"Bearer {make_test_token_for_user(TEST_USER_ID)}"}


def get_test_admin_headers() -> dict:
    """Get authorization headers for the test admin."""
    return {
        "Authorization": f"Bearer {make_test_token_for_user(TEST_ADMIN_ID, role='super_admin')}"
    }


# ==========================================
# PYTEST FIXTURES
# ==========================================

# Pytest fixtures are only available when pytest is installed
# Use: from tests.fixtures import test_user_id

try:
    import pytest

    @pytest.fixture
    def test_user_id() -> str:
        """Fixture providing the test user ID."""
        return TEST_USER_ID

    @pytest.fixture
    def test_admin_id() -> str:
        """Fixture providing the test admin ID."""
        return TEST_ADMIN_ID

    @pytest.fixture
    def test_user_headers() -> dict:
        """Fixture providing authorization headers for test user."""
        return get_test_user_headers()

    @pytest.fixture
    def test_admin_headers() -> dict:
        """Fixture providing authorization headers for test admin."""
        return get_test_admin_headers()

except ImportError:
    # pytest not installed, fixtures not available
    test_user_id = None
    test_admin_id = None
    test_user_headers = None
    test_admin_headers = None


# ==========================================
# SUPABASE AUTH HELPERS
# ==========================================


async def sign_in_test_user(supabase_client) -> dict:
    """
    Sign in the test user using Supabase auth.

    Args:
        supabase_client: An initialized Supabase client

    Returns:
        The auth response from Supabase
    """
    return await supabase_client.auth.sign_in_with_password(
        {
            "email": TEST_USER_EMAIL,
            "password": TEST_USER_PASSWORD,
        }
    )


async def sign_in_test_admin(supabase_client) -> dict:
    """
    Sign in the test admin using Supabase auth.

    Args:
        supabase_client: An initialized Supabase client

    Returns:
        The auth response from Supabase
    """
    return await supabase_client.auth.sign_in_with_password(
        {
            "email": TEST_ADMIN_EMAIL,
            "password": TEST_ADMIN_PASSWORD,
        }
    )


# ==========================================
# VERIFICATION HELPERS
# ==========================================


def verify_test_accounts_exist(db_query_func) -> bool:
    """
    Verify that both test accounts exist in the database.

    Args:
        db_query_func: A function that executes a SQL query and returns results

    Returns:
        True if both accounts exist, False otherwise
    """
    try:
        user_result = db_query_func("SELECT id FROM public.users WHERE id = %s", (TEST_USER_ID,))
        admin_result = db_query_func("SELECT id FROM public.users WHERE id = %s", (TEST_ADMIN_ID,))
        admin_role_result = db_query_func(
            "SELECT role FROM public.admin_roles WHERE user_id = %s", (TEST_ADMIN_ID,)
        )

        return bool(
            user_result
            and admin_result
            and admin_role_result
            and admin_role_result[0]["role"] == "super_admin"
        )
    except Exception:
        return False


if __name__ == "__main__":
    # Quick verification script
    print("=" * 60)
    print("PERMANENT TEST ACCOUNTS")
    print("=" * 60)
    print()
    print("Test User:")
    print(f"  ID:       {TEST_USER_ID}")
    print(f"  Email:    {TEST_USER_EMAIL}")
    print(f"  Password: {TEST_USER_PASSWORD}")
    print()
    print("Test Admin:")
    print(f"  ID:       {TEST_ADMIN_ID}")
    print(f"  Email:    {TEST_ADMIN_EMAIL}")
    print(f"  Password: {TEST_ADMIN_PASSWORD}")
    print(f"  Role:     {TEST_ADMIN_ROLE}")
    print()
    print("=" * 60)
    print("Use in your automation tests!")
    print("=" * 60)

File: tests/fixtures/test-accounts.ts
/**
 * Permanent Test Accounts for Automation Testing
 * 
 * These accounts are created by migration:
 * supabase/migrations/20260303_create_permanent_test_accounts.sql
 * 
 * They will always exist after running migrations and can be used
 * for integration tests, e2e tests, and automation.
 */

// ==========================================
// TEST ACCOUNT CONSTANTS
// ==========================================

/** Test User Account (Regular User) */
export const TEST_USER = {
  id: 'a1111111-1111-1111-1111-111111111111',
  email: 'test.user@hanachan.test',
  password: 'TestPassword123!',
  displayName: 'Test User',
} as const;

/** Test Admin Account (Super Admin) */
export const TEST_ADMIN = {
  id: 'b2222222-2222-2222-2222-222222222222',
  email: 'test.admin@hanachan.test',
  password: 'AdminPassword123!',
  displayName: 'Test Admin',
  role: 'super_admin',
} as const;

/** All test accounts for easy iteration */
export const ALL_TEST_ACCOUNTS = [TEST_USER, TEST_ADMIN] as const;

// ==========================================
// TYPE DEFINITIONS
// ==========================================

export interface TestAccount {
  id: string;
  email: string;
  password: string;
  displayName: string;
  role?: string;
}

// ==========================================
// SUPABASE AUTH HELPERS
// ==========================================

import { createClient, SupabaseClient } from '@supabase/supabase-js';

/**
 * Sign in the test user using Supabase auth
 */
export async function signInTestUser(supabase: SupabaseClient) {
  return await supabase.auth.signInWithPassword({
    email: TEST_USER.email,
    password: TEST_USER.password,
  });
}

/**
 * Sign in the test admin using Supabase auth
 */
export async function signInTestAdmin(supabase: SupabaseClient) {
  return await supabase.auth.signInWithPassword({
    email: TEST_ADMIN.email,
    password: TEST_ADMIN.password,
  });
}

/**
 * Get authenticated Supabase client for test user
 */
export async function getAuthenticatedUserClient(supabaseUrl: string, supabaseKey: string): Promise<SupabaseClient> {
  const supabase = createClient(supabaseUrl, supabaseKey);
  const { error } = await signInTestUser(supabase);
  if (error) throw error;
  return supabase;
}

/**
 * Get authenticated Supabase client for test admin
 */
export async function getAuthenticatedAdminClient(supabaseUrl: string, supabaseKey: string): Promise<SupabaseClient> {
  const supabase = createClient(supabaseUrl, supabaseKey);
  const { error } = await signInTestAdmin(supabase);
  if (error) throw error;
  return supabase;
}

// ==========================================
// PLAYWRIGHT / E2E HELPERS
// ==========================================

/**
 * Login as test user in a Playwright page
 */
export async function loginAsTestUser(page: any, baseUrl: string): Promise<void> {
  await page.goto(`${baseUrl}/login`);
  await page.fill('[data-testid="email-input"]', TEST_USER.email);
  await page.fill('[data-testid="password-input"]', TEST_USER.password);
  await page.click('[data-testid="login-button"]');
  await page.waitForNavigation();
}

/**
 * Login as test admin in a Playwright page
 */
export async function loginAsTestAdmin(page: any, baseUrl: string): Promise<void> {
  await page.goto(`${baseUrl}/login`);
  await page.fill('[data-testid="email-input"]', TEST_ADMIN.email);
  await page.fill('[data-testid="password-input"]', TEST_ADMIN.password);
  await page.click('[data-testid="login-button"]');
  await page.waitForNavigation();
}

// ==========================================
// MOCK DATA HELPERS
// ==========================================

/**
 * Get mock JWT token payload for test user
 */
export function getTestUserJwtPayload() {
  return {
    sub: TEST_USER.id,
    email: TEST_USER.email,
    role: 'authenticated',
    aud: 'authenticated',
  };
}

/**
 * Get mock JWT token payload for test admin
 */
export function getTestAdminJwtPayload() {
  return {
    sub: TEST_ADMIN.id,
    email: TEST_ADMIN.email,
    role: 'super_admin',
    aud: 'authenticated',
  };
}

// ==========================================
// DEBUG HELPERS
// ==========================================

/**
 * Print test account info to console
 */
export function printTestAccounts(): void {
  // eslint-disable-next-line no-console
  console.log('='.repeat(60));
  // eslint-disable-next-line no-console
  console.log('PERMANENT TEST ACCOUNTS');
  // eslint-disable-next-line no-console
  console.log('='.repeat(60));
  // eslint-disable-next-line no-console
  console.log();
  // eslint-disable-next-line no-console
  console.log('Test User:');
  // eslint-disable-next-line no-console
  console.log(`  ID:       ${TEST_USER.id}`);
  // eslint-disable-next-line no-console
  console.log(`  Email:    ${TEST_USER.email}`);
  // eslint-disable-next-line no-console
  console.log(`  Password: ${TEST_USER.password}`);
  // eslint-disable-next-line no-console
  console.log();
  // eslint-disable-next-line no-console
  console.log('Test Admin:');
  // eslint-disable-next-line no-console
  console.log(`  ID:       ${TEST_ADMIN.id}`);
  // eslint-disable-next-line no-console
  console.log(`  Email:    ${TEST_ADMIN.email}`);
  // eslint-disable-next-line no-console
  console.log(`  Password: ${TEST_ADMIN.password}`);
  // eslint-disable-next-line no-console
  console.log(`  Role:     ${TEST_ADMIN.role}`);
  // eslint-disable-next-line no-console
  console.log();
  // eslint-disable-next-line no-console
  console.log('='.repeat(60));
}
